### PR TITLE
User validation

### PR DESCRIPTION
**Pull request by diPhantxm**
Simple user name validation by comparing names in config and psql query

### CODE REVIEW

**Comment by reshke** [comment]
Connecting to console admin under user name, which does not configured in rules is not itself a big issue. 
But if changes affects only psql (not grpc) interface, thats maybe worth it ok. 

I would rather think of auth mechanism for console interface, as well as for coordinator-related interactions (through both gprc and postgres-proto ifaces). Storing auth token in config and requesting it during startup/gRPC authentication would be sufficient IMO

**Comment by diPhantxm** [comment]
Made it like this. It now supports password but there is as issue. After requesting password from user it sends ErrorResponse and auth failed, but after user entered password everything is fine
```
LOG: 2023/01/20 13:15:51.541664 init client connection with ssl: false
DEBUG1: 2023/01/20 13:15:51.541933 client param application_name psql
DEBUG1: 2023/01/20 13:15:51.541938 client param client_encoding UTF8
DEBUG1: 2023/01/20 13:15:51.541940 client param user news
DEBUG1: 2023/01/20 13:15:51.541942 client param database news
DEBUG4: 2023/01/20 13:15:51.541950 allocate route {news news}
LOG: 2023/01/20 13:15:51.541955 processing auth for news news
DEBUG3: 2023/01/20 13:15:51.542122 sending *pgproto3.ErrorResponse to client
DEBUG3: 2023/01/20 13:15:51.542149 sending *pgproto3.ErrorResponse to client
ERROR: 2023/01/20 13:15:51.542162 user news news auth failed
LOG: 2023/01/20 13:15:53.692890 init client connection with ssl: false
DEBUG1: 2023/01/20 13:15:53.693291 client param user news
DEBUG1: 2023/01/20 13:15:53.693327 client param database news
DEBUG1: 2023/01/20 13:15:53.693344 client param application_name psql
DEBUG1: 2023/01/20 13:15:53.693360 client param client_encoding UTF8
INFO: 2023/01/20 13:15:53.693392 match route {news news}
LOG: 2023/01/20 13:15:53.693409 processing auth for news news
DEBUG3: 2023/01/20 13:15:53.693722 sending *pgproto3.AuthenticationOk to client
LOG: 2023/01/20 13:15:53.693807 acquire conn to sh1
LOG: 2023/01/20 13:15:53.693827 acquire new connection to ???
DEBUG3: 2023/01/20 13:15:53.694211 init new postgresql instance connection to ???
DEBUG2: 2023/01/20 13:15:53.772614 Auth type proc *pgproto3.AuthenticationMD5Password
DEBUG1: 2023/01/20 13:15:53.772655 sending auth package b4b71a9daa3c7198efcb85605342d038 plain passwd Password.1
DEBUG2: 2023/01/20 13:15:53.789237 Auth type proc *pgproto3.AuthenticationOk
DEBUG1: 2023/01/20 13:15:53.789325 ignored backend key data 150680245 804504791
DEBUG3: 2023/01/20 13:15:53.789358 sending *pgproto3.ParameterStatus to client
DEBUG3: 2023/01/20 13:15:53.789451 sending *pgproto3.ParameterStatus to client
DEBUG3: 2023/01/20 13:15:53.789508 sending *pgproto3.ParameterStatus to client
DEBUG3: 2023/01/20 13:15:53.789564 sending *pgproto3.ParameterStatus to client
DEBUG3: 2023/01/20 13:15:53.789616 sending *pgproto3.ParameterStatus to client
DEBUG3: 2023/01/20 13:15:53.789667 sending *pgproto3.ParameterStatus to client
DEBUG3: 2023/01/20 13:15:53.789715 sending *pgproto3.ParameterStatus to client
DEBUG3: 2023/01/20 13:15:53.789765 sending *pgproto3.ParameterStatus to client
DEBUG3: 2023/01/20 13:15:53.789814 sending *pgproto3.ParameterStatus to client
DEBUG3: 2023/01/20 13:15:53.789855 sending *pgproto3.ParameterStatus to client
DEBUG3: 2023/01/20 13:15:53.789908 sending *pgproto3.ParameterStatus to client
DEBUG3: 2023/01/20 13:15:53.789959 sending *pgproto3.ParameterStatus to client
DEBUG3: 2023/01/20 13:15:53.790009 sending *pgproto3.ReadyForQuery to client
LOG: 2023/01/20 13:15:53.790061 client auth OK
DEBUG3: 2023/01/20 13:15:53.790085 sending *pgproto3.AuthenticationOk to client
DEBUG3: 2023/01/20 13:15:53.790124 sending *pgproto3.ParameterStatus to client
DEBUG3: 2023/01/20 13:15:53.790154 sending *pgproto3.ParameterStatus to client
DEBUG3: 2023/01/20 13:15:53.790181 sending *pgproto3.NoticeResponse to client
DEBUG3: 2023/01/20 13:15:53.790236 sending *pgproto3.ReadyForQuery to client
LOG: 2023/01/20 13:15:53.790265 console.ProcClient start
```

```
diphantxm@diphantxm-nux:~/Desktop$ psql "sslmode=disable host=localhost dbname=news user=news port=7432"
Password for user news: 
psql (12.12 (Ubuntu 12.12-0ubuntu0.20.04.1), server 14.6 (Ubuntu 14.6-201-yandex.52662.649e571508))
WARNING: psql major version 12, server major version 14.
         Some psql features might not work.
Type "help" for help.

news=> 

```

**Comment by reshke** [comment]
It seems like we either need to rewrite Preroute logic to be more generic, and thus work for console-db case or write separate Preroute implementation for console-db. For example,  you can see from log messages, that router tries to acquire shard contentions while processing console(admin) session, which is absolutely unneeded.  

Personally, i would prefer separate Preroute implementation for console-db for now.

**Comment by diPhantxm** [comment]
I can do smth like this 
```
func (r *InstanceImpl) servAdm(ctx context.Context, conn net.Conn) error {
	cl := client.NewPsqlClient(conn)
	if err := cl.Init(r.frTLS); err != nil {
		return err
	}

	if err := r.validate(cl); err != nil {
		return err
	}

	return r.AdmConsole.Serve(ctx, cl)
}

func (r *InstanceImpl) validate(cl *client.PsqlClient) error {
	for _, rule := range r.RuleRouter.Config().FrontendRules {
		if cl.Usr() == rule.Usr {
			cl.AssignRule(rule)
			return conn.AuthFrontend(cl, cl.Rule().AuthRule)
		}
	}

	return fmt.Errorf("no configuration rule for user %s", cl.Usr())
}
```
As far as I understand, it does the same thing but without any backend part. Not sure if it's needed. Is it ok?

**Comment by reshke** [comment]
yeah, sure, try to use map instead of linear search, it should be mapping from client username to rule somewhere in rulerouter

File: router/pkg/rulerouter/rulerouter.go [hunk 1]
```diff
@@ -192,6 +194,37 @@ func (r *RuleRouterImpl) PreRoute(conn net.Conn) (rclient.RouterClient, error) {
...
+	if err := cl.Init(r.tlsconfig); err != nil {
+		return nil, err
+	}
+
+	key := *route.NewRouteKey(cl.Usr(), cl.DB())
+	frRule, err := r.rmgr.MatchKeyFrontend(key)
+	if err != nil {
+		for _, msg := range []pgproto3.BackendMessage{
+			&pgproto3.ErrorResponse{
+				Message: err.Error(),
```
Comment (reshke) [inline router/pkg/rulerouter/rulerouter.go:208]: cl.ReplyErrMsg("failed to make route failure response")

**Review by reshke — APPROVED** [review]

**Pull request merged**